[wallet] Remove vchDefaultKey and have better first run detection

achow101 opened the pull request:
Removes vchDefaultKey which was only used for first run detection. Improves wallet first run detection by checking to see if any keys were read from the database.

This also fixes a (rather contrived) case where an encrypted non-HD wallet has corruption such that the default key is no longer valid and is loaded into a Core version that supports HD wallets. This causes a runtime exception since a new hd master key is generated as the software believes the wallet file is newly created but cannot add the generated key to the wallet since it is encrypted. I was only able to replicate this error by creating a non-hd wallet, encrypting it, then editing the wallet using `db_dump` and `db_load` before loading the wallet with hd enabled. This problem has been reported by [two](https://bitcointalk.org/index.php?topic=1993244.0) [users](https://bitcointalk.org/index.php?topic=1746976.msg17511261#msg17511261) so it is something that can happen, although that raises the question of "what corrupted the default key".

~P.S. I don't know what's up with the whitespace changes. I think my text editor is doing something stupid but I don't think those are important enough to attempt undoing them.~ Undid those

achow101 committed:
Remove vchDefaultKey and have better first run detection

Removes vchDefaultKey which was only used for first run detection.
Improves wallet first run detection by checking to see if any keys
were read from the database.

This will now also check for a valid defaultkey for backwards
compatibility reasons and to check for any corruption.

Keys will stil be generated on the first one, but there won't be
any shown in the address book as was previously done.

gmaxwell commented:
So you still need to write a default key unless we're going to bump the wallet version again for this, because otherwise you'll make it backwards incompatible with software without this patch.

Otherwise, Concept ACK!

achow101 commented:
How about bump the wallet version again? If there are any other wallet changes that need a version bump, we could do them all at the same time.

gmaxwell commented:
I would be fine with that but I think it might be easier to get this bugfix merged without it.

TheBlueMatt commented:
Sounds like maybe something that we should look into for 15, then, though I'd really like to look into the source of the corruption here, rather than patching over it.

achow101 commented:
@gmaxwell I added `WriteDefaultKey` back in so that a default key will be written to maintain backwards compatibility. The key will not be read in.

gmaxwell reviewed (approved):
utACK

TheBlueMatt commented:
I really dont think this is the right way to fix this issue. Can we instead ensure that, if the wallet has keys but vchDefaultKey is invalid, an error is printed like "your wallet is corrupted, seems you disk is silently corrupting things", and then make sure -salvagewallet is able to properly recover from this state?

achow101 commented:
I have updated this with a few changes.

I removed the defaultkey writing again. New wallets will no longer have a default key. If a defaultkey is found in the wallet, it must be valid, otherwise it will exit with a wallet corruption error.

It turns out that `-salvagewallet` may have been the cause of all of these issues. `-salvagewallet` apparently does not write a default key to the wallet, which with an encrypted wallet and with HD enabled, will cause the aforementioned runtime exception and crash. If a user does have a corrupted default key, then `-salvagewallet` will remove it, as it did in the past.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -3958,11 +3947,13 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
+
+        // Generate an address for the address book and to fill the keypool
+        CPubKey key;
+        if (walletInstance->GetKeyFromPool(key, false)) {
```
Can you call `walletInstance->TopUpKeyPool();` here instead of this entire section? No need to waste a key and an address book entry for this.

achow101 replied:
Done

sipa commented on src/wallet/walletdb.cpp:
```diff
@@ -452,7 +447,13 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+            ssKey >> vchPubKey;
+            if (!vchPubKey.IsValid())
+            {
+                strErr = "Error reading wallet database: Default Key corrupt";
```
"defaultkey" is not part of `IsKeyType`, so this failure will just be a non-critical warning upon loading. I assume that's intentional?

achow101 replied:
It should be a critical warning and exit with an error. I have fixed this.

sipa commented on src/wallet/walletdb.cpp:
```diff
@@ -452,7 +447,13 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+            CPubKey vchPubKey;
+            ssKey >> vchPubKey;
+            if (!vchPubKey.IsValid())
+            {
```
Code style nit: `{` on the same like as `if`.

achow101 replied:
Done

MarcoFalke commented:
AssertionError in wallet-hd:

```py
AssertionError: not(m/0'/0'/0' == m/0'/0'/1')

achow101 commented:
Can this be tagged for 0.15.0? It's a bug fix for a problem with `-salvagewallet` which will result in further corruption and unusability of people's wallets.

sipa commented:
Tagged for 0.15, as it fixes a bug when using `-savagewallet` with encrypted wallets.

TheBlueMatt commented:
Because one of the two cases of corruption in the wild was that the default key was simply deleted, can you also add a if(wallet version < x && !have_default_key)... in the load?

On August 4, 2017 2:17:43 AM EDT, MarcoFalke <notifications@github.com> wrote:
>AssertionError in wallet-hd:
>
>```py
>AssertionError: not(m/0'/0'/0' == m/0'/0'/1')
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/10952#issuecomment-320166041

achow101 commented:
@TheBlueMatt What for?

TheBlueMatt commented:
@achow101 because its a trivial check and its a type of corruption we've seen in wallets on the wild...it costs nothing to add a check for it.

sipa commented:
@TheBlueMatt The current code requires that if a default key is present, it must be valid. Do you think an extra check is needed to verify that it hasn't been accidentally deleted?

TheBlueMatt commented:
Considering that we've seen it in the wild and its clear indication that
there is something wrong with your hardware/wallet storage, I'd say it
would be prudent.

On 08/06/17 21:16, Pieter Wuille wrote:
> @TheBlueMatt <https://github.com/thebluematt> The current code requires
> that if a default key is present, it must be valid. Do you think an
> extra check is needed to verify that it hasn't been accidentally deleted?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/10952#issuecomment-320546207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnoHoggbwbRS_yma0eQ6gmvVwaxu4YJks5sVmV0gaJpZM4OnMtA>.
>

achow101 commented:
@TheBlueMatt I meant what do you want it to check for and error on? That below a certain version number you must have a valid default key? If that is the case, that isn't possible since this PR does not change the version number, thus all wallets will have to have a valid default key.

Furthermore, I believe the corruption that we are seeing in the wild is a result of `-savagewallet` not passing the default key through to the new wallet file, not because of hardware corruption or the like.

TheBlueMatt commented:
Ahh, OK, I missed the salvagewallet-possible-culprit. PR looks good
as-is to me.

On 08/06/17 21:36, Andrew Chow wrote:
> @TheBlueMatt <https://github.com/thebluematt> I meant what do you want
> it to check for and error on? That below a certain version number you
> must have a valid default key? If that is the case, that isn't possible
> since this PR does not change the version number, thus all wallets will
> have to have a valid default key.
> 
> Furthermore, I believe the corruption that we are seeing in the wild is
> a result of |-savagewallet| not passing the default key through to the
> new wallet file, not because of hardware corruption or the like.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/10952#issuecomment-320548013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnoHh7moN-GrfZywt97DqoUu2eT-Mc6ks5sVmoVgaJpZM4OnMtA>.
>

laanwj commented:
Nit: please also remove the `"defaultkey"` parsing in `walletdb.cpp` `ReadKeyValue`

achow101 commented:
@laanwj That was left in due address @TheBlueMatt's concerns about warning users about possible corruption. It was modified so that the default key must either exist and be valid or not exist at all. The reason for this is that if it exists and is not valid, then corruption has occurred and the user will be warned accordingly.

laanwj commented:
That's true, could have used a comment, it now looks a bit silly just throwing away the result. But I agree.

laanwj commented:
I tested this with an old wallet and got "Error: Error loading wallet.dat: Wallet corrupted", will check why.

laanwj commented on src/wallet/walletdb.cpp:
```diff
@@ -452,7 +447,12 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
         {
-            ssValue >> pwallet->vchDefaultKey;
+            CPubKey vchPubKey;
+            ssKey >> vchPubKey;
```
This should be `ssValue >> vchPubKey`. Otherwise this will fail for every wallet with a default key record!

achow101 replied:
Oops. Fixed.

jnewbery commented:
Intuitively for me, the logic for determining whether this is a first run should remain in the `CWallet::LoadWallet()` function. It makes sense for me that `CWalletDB` remains as a modifier object for the wallet, and simply reads/writes data to the wallet database. Is it possible to leave the first_run logic in `CWallet::LoadWallet()` (by checking that `mapKeys` is empty or similar)?

I'm not all that familiar with wallet code, so I may be wrong about this. I'll defer to those who have a better understanding of the wallet model.

I also don't think this necessarily needs to go into v0.15 if we don't have time to fully review it. The problems with `salvagewallet` are long-standing and not made any worse in v0.15.

achow101 commented:
>  Is it possible to leave the first_run logic in CWallet::LoadWallet() (by checking that mapKeys is empty or similar)?

That would probably be better to do.

> I also don't think this necessarily needs to go into v0.15 if we don't have time to fully review it. The problems with `salvagewallet` are long-standing and not made any worse in v0.15.

0.13 broke `salvagewallet` with HD wallets. Anyone who uses `salvagewallet` on an encrypted wallet is currently going to run into problems just starting their wallet in the first place.

jnewbery commented:
> 0.13 broke salvagewallet with HD wallets.

To be clear, I'm not arguing against this PR - it's obviously good and we should do it. My point is that we don't need to rush it for v0.15. The number of users who haven't yet upgraded to v0.13, but will upgrade to v0.15 before we release v0.15.1 is going to be ~ zero. Not including the fix in v0.15 doesn't make the problem any worse.

laanwj commented:
> Intuitively for me, the logic for determining whether this is a first run should remain in the CWallet::LoadWallet() function.

I thought about this too and came to the opposite conclusion: wouldn't the database layer be the foremost authority on whether it just created a new database file, or opened an existing one? I think this PR goes the right direction.

jnewbery commented:
I guess it depends on what we mean by 'first run'. If it's 'the first time we create a file', then yes I agree that the database layer is the authority. If it's 'a wallet without any of these objects', then I would have thought the right place would be at the wallet layer.

But that's just my intuition. You know this code better than I do, so if you think this is the right direction, I won't object.

laanwj commented:
> I guess it depends on what we mean by 'first run'. If it's 'the first time we create a file', then yes I agree that the database layer is the authority. If it's 'a wallet without any of these objects', then I would have thought the right place would be at the wallet layer.

CWalletDB is not just the low-level database though - it also manages the format. After all it iterates over application-specific key/value pairs in `CWalletDB::LoadWallet`. So any "application format" checking belongs there too, IMO. Counting "a wallet without any of these objects" belongs naturally in the place where all the objects are iterated over.

`wallet/db.cpp` has the low-level, application-independent database handling. If you'd argue that that is an abstraction level too low to put this, I'd agree.

I also can't help that all these classes and objects are named confusingly.

laanwj commented:
>To be clear, I'm not arguing against this PR - it's obviously good and we should do it. My point is that we don't need to rush it for v0.15. 

I'd prefer to drop it for 0.15.0 too, to be honest.

achow101 commented:
I changed the detection back to `CWallet::LoadWallet()` since I think it is more logical to be there. That was my original plan, I just couldn't figure out how since I forgot that it has access to `mapKeys`, etc.

jnewbery commented:
Looks good. Tested ACK b8aa9729a5dbeb883857b14918aa6c6d2b71eea4 (although needs rebase)

jonasschnelli commented:
Needs rebase.

achow101 commented:
rebased

jnewbery commented:
utACK 1a7bc5aac60a37b6f9f0632546467d716a3fd5e4

sipa commented:
There seems to be a problem in `keypool_topup.py`.

achow101 commented:
Fixed that I think. It's an off-by-one since default key is no longer created and consuming a key from the keypool.

laanwj commented:
> It's an off-by-one since default key is no longer created and consuming a key from the keypool.

Finally :-)

jonasschnelli commented:
Great to see the default key vanish...
utACK e53615b443894fb96d5eb885b3812776c1b1033b

laanwj commented:
Removing 0.15.1 milestone, as it can create problems at the moment when opening a new wallet with older versions.
